New tutorial for ROS 2 nodes (#343)
ros2/ros2_documentation: source/Tutorials/UnderstandingROS2Nodes.rst

.. _ROS2Nodes:

Understanding ROS 2 nodes
=========================

**Goal:** Learn about the function of nodes in ROS 2, and the tools to interact with them.

**Tutorial level:** Beginner

**Time:** 10 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

1 The ROS 2 graph
^^^^^^^^^^^^^^^^^

Over the next few tutorials, you will learn about a series of core ROS 2 concepts that make up what is referred to as the “ROS (2) graph”.

The ROS graph is the collection of ROS 2 elements processing data together at one time.

> It encompasses all executables and the connections between them, if you were to map them all out and `visualize them <https://www.researchgate.net/figure/An-example-of-a-ROS-program-It-is-a-graph-where-a-node-is-a-process-providing-or_fig1_319138847>`__.

Reviewer: ```suggestion
It encompasses all executables and the connections between them; if you were to map them all out and `visualize them <https://www.researchgate.net/figure/An-example-of-a-ROS-program-It-is-a-graph-where-a-node-is-a-process-providing-or_fig1_319138847>`__.
```

Reviewer: As you mention in your original comment, this links outside of our organization.  I'd really prefer not to do this, as if that site goes down or changes the URL on us, we have no control over that.  The good news is that it should be pretty easy to recreate something like that using [Graphviz](https://renenyffenegger.ch/notes/tools/Graphviz/examples/index), so I'd suggest doing that instead.

2 Nodes in ROS 2
^^^^^^^^^^^^^^^^

Nodes are the fundamental communication type in the ROS graph.
A node represents a single modular process.
A robot system would be comprised of many nodes; for example, one controlling wheel motors, another controlling a laser range-finder, and so on.
In ROS 2, a single executable (C++ program, Python program, etc.) can contain multiple nodes.

Prerequisites
-------------

.. The previous tutorial, :ref:`Turtlesim2`, shows you how to install the turtlesim package used here.

> As always, don’t forget to source ROS 2 in **every new terminal you open**.

Author: Good point, I've been thinking about this, especially since it's listed in the prereqs of every tutorial.

I'm thinking about linking to the "Configuring your environment" tutorial every time it says this.

> As always, don’t forget to source ROS 2 in **every new terminal you open**.

Author: Since I have to do that for every tutorial, I think I'll do it in one fell swoop once everything's merged into staging (unless that's bad practice?)

Tasks
-----

1 ros2 run
^^^^^^^^^^

The command ``ros2 run`` launches an executable from a package.

.. code-block:: bash

    ros2 run <package_name> <executable_name>

To run turtlesim, open a new terminal, and enter the following command:

.. code-block:: bash

    ros2 run turtlesim turtlesim_node

The turtlesim window will open, as you saw in the previous tutorial.

Here, the package name is ``turtlesim`` and the executable name is ``turtlesim_node``.

We still don’t know the node name, however.
You can find node names by using ``ros2 node list``

2 ros2 node list
^^^^^^^^^^^^^^^^

``ros2 node list`` will show you the names of all running nodes.
This is especially useful when you want to interact with a node, or when you have a system running many nodes and need to keep track of them.

Open a new terminal while turtlesim is still running in the other one, and enter the following command:

.. code-block:: bash

    ros2 node list

> The terminal will return the node name ``/turtlesim``

Reviewer: I think you should do a `codeblock` here to show the node name, so it is consistent with the list of nodes that you print out below.

Open another new terminal and start the teleop node with the command:

.. code-block:: bash

    ros2 run turtlesim turtle_teleop_key

Here, we are searching the ``turtlesim`` package again, this time for the executable named ``turtle_teleop_key``.

Return to the terminal where you ran ``ros2 node list`` and run it again.

> You will now see the names of two active nodes, ``/turtlesim`` and ``/teleop_turtle``.

Reviewer: Again, I think a `codeblock` here to show the node names will be more consistent than below.

2.1 Remapping
~~~~~~~~~~~~~

> Remapping allows you to reassign default node properties from the command line.

Reviewer: This seems like a reasonable way to describe it.  You may want to also add specifics here, saying that you can remap things like the node name, the topic names, the service names, etc.

.. _ROS2Nodes:

Understanding ROS 2 nodes
=========================

**Goal:** Learn about the function of nodes in ROS 2, and the tools to interact with them.

**Tutorial level:** Beginner

**Time:** 10 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

1 The ROS 2 graph
^^^^^^^^^^^^^^^^^

Over the next few tutorials, you will learn about a series of core ROS 2 concepts that make up what is referred to as the “ROS (2) graph”.

The ROS graph is the collection of ROS 2 elements processing data together at one time.
It encompasses all executables and the connections between them if you were to map them all out and visualize them.

2 Nodes in ROS 2
^^^^^^^^^^^^^^^^

Nodes are the fundamental communication type in the ROS graph.

> A node represents a single modular process.

Reviewer: The thing is, nodes aren't a communication type at all.  If you just have a node, no communication is possible.  A node is the framework on which you hang different communication types, like topics, services, actions, etc, so I think this explanation needs to reflect that.

Reviewer 2: I would say the word "component"?  

> Nodes are a fundamental component in the ROS graph.  It provides the necessary framework for connecting to topics, services, actions, and parameters in the rest of the ROS graph.

Author: How about what a node actually is, like "a container for the functionality of a single process, which contains the details of all the other communication types in ROS 2" or something like that (if that's a correct statement)?

@dirk-thomas I've been using "components" to describe pretty much everything in this part of the tutorial series: nodes, services, topics, actions, etc. What should I say instead?

Reviewer: > How about what a node actually is, like "a container for the functionality of a single process, which contains the details of all the other communication types in ROS 2" or something like that (if that's a correct statement)?

So the container part is correct, but it's not for a single process.  That is, you can have multiple nodes per-process.

When it comes down to it, I actually am not entirely sure why we have nodes in ROS 2.  I have a vague feeling we really want them, but I'm just not sure why.  Does anyone else have something more concrete to say about nodes here?

Author: By process I meant that a single node serves a modular purpose, does that make sense?

>a container of functionality that serves a single, modular purpose, and contains the details of all the other communication types in ROS 2

This is getting really convoluted haha

Reviewer: I agree it's convoluted, but I think this is an important part of this article.  To be more clear, I think it might make sense to combine this whole paragraph together into something like this:

```
Each node in ROS should be responsible for a single, module purpose (e.g. one node for controlling wheel motors, one node for controlling a laser range-finder, etc).  Each node can send and receive data to other nodes via topics, services, actions, or parameters.  A full robotic system would be comprised of many of these nodes working in concert.  In ROS 2, a single executable (C++ program, Python program, etc.) can contain one or more nodes.
```

Author: That's great @clalancette, thanks!

A robot system would be comprised of many nodes; for example, one controlling wheel motors, another controlling a laser range-finder, and so on.

> In ROS 2, a single executable (C++ program, Python program, etc.) can contain multiple nodes.

Reviewer: ```suggestion
In ROS 2, a single executable (C++ program, Python program, etc.) can contain one or more nodes.
```

Prerequisites
-------------

.. The previous tutorial, :ref:`Turtlesim2`, shows you how to install the turtlesim package used here.

As always, don’t forget to source ROS 2 in **every new terminal you open**.

Tasks
-----

1 ros2 run
^^^^^^^^^^

> The command ``ros2 run`` launches an executable from a package.

Reviewer: It doesn't seem like this explanation is needed here, since this is specifically talking bout nodes, not packages.  Isn't there an earlier tutorial explaining packages and executables and such?

Reviewer 2: I think that I had suggested in my conversation with @maryaB-osr that we draw the distinction that `ros2 run` is no longer starting "nodes" in ROS 2, since a single process can have multiple nodes.  The 1-to-1 mapping between executable and node isn't quite the same, so I think it makes sense to be consistent across the tutorials.

Author: @clalancette the [previous tutorial](https://github.com/ros2/ros2_documentation/blob/staging/source/Tutorials/Turtlesim/IntroducingTurtlesim.rst) starts by introducing ``turtlesim`` as a package, and also immediately tells readers how to see the executables inside turtlesim. 

True, it doesn't come up anywhere else but the structure of ``ros2 run`` is sort of explained along the way in this tutorial.

But like @mjcarroll said, I want to start the reader using the language `packages contain > executables contain> nodes` because the distinction does come up in the future, for example the nodes included in [the launch file tutorial](https://github.com/ros2/ros2_documentation/pull/353/files)

Do you think:
```
The command ``ros2 run`` launches a node from an executable inside a package.
```
Is better?

Reviewer: I didn't explain myself that well, but I was more asking if the topic belongs at all in this tutorial, or belongs in a more dedicated place.  I don't think it warrants the confusion about nodes/executables at this stage, and it's a minor point, so I think we can just leave the existing wording.

      /my_turtle/get_parameters: rcl_interfaces/srv/GetParameters
      /my_turtle/list_parameters: rcl_interfaces/srv/ListParameters
      /my_turtle/set_parameters: rcl_interfaces/srv/SetParameters
      /my_turtle/set_parameters_atomically: rcl_interfaces/srv/SetParametersAtomically
      /reset: std_srvs/srv/Empty
      /spawn: turtlesim/srv/Spawn
      /turtle1/set_pen: turtlesim/srv/SetPen
      /turtle1/teleport_absolute: turtlesim/srv/TeleportAbsolute
      /turtle1/teleport_relative: turtlesim/srv/TeleportRelative

.. todo: (change after actions are added)

Now try running the same command on the ``/teleop_turtle`` node, and see how its connections differ from ``my_turtle``.

You will learn more about ROS graph connection concepts in the upcoming tutorials.

Summary
-------

A node is a fundamental ROS 2 element that serves a single, modular purpose in a robotics system.

In this tutorial, you utilized nodes created from the ``turtlesim`` package by running the executables ``turtlesim_node`` and ``turtle_teleop_key``.

You learned how to use ``ros2 node list`` to discover active node names and ``ros2 node info`` to introspect on a single node.

> Hopefully now you can understand how these tools would be vital to understanding the flow of data in a complex, real-world robot system.

Reviewer: I'm not sure I like the use of the world "Hopefully" here.  How about: 

```suggestion
These tools are vital to understanding the flow of data in a complex, real-world robot system.
```
